### PR TITLE
- restore freshon search plugin from 8636793fe6532fee086ec6664eefb844…

### PR DESCRIPTION
**Pull request by Jesus-21**
### Motivation for changes:
Freshon.tv search plugin is documented but was not available.
I guess it has not been ported when the python3 compatibility has been done.

### Detailed changes:

- I restored the plugin from commit 8636793fe6532fee086ec6664eefb844
- moved the file from `flexget/plugins/search_freshon.py` to `flexget/plugins/sites/freshon.py` 
- As advised in https://flexget.com/Developers#python23-compatibility I added from the original file
```python
from builtins import *
```
- Only change from original plugin is that I removed `verify=False` the post request

### Config usage if relevant (new plugin or updated schema):
```
tv_search_freshon:
    discover:
      what:
         - trakt_list:
             username: xxxxxxx
             password: xxxxxxx
             list: somelist
             type: shows
          from:
            - freshon:
                username: xxxxxx
                password: xxxxxx
                passkey: xxxxx
```
Note: `passkey` is not present in documentation but is/was mandatory (cf. https://flexget.com/Searches/freshon )

### CODE REVIEW

File: flexget/plugins/sites/freshon.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+                        seeds = int(res.find('td', {'class': 'table_seeders'}).find('span').text.strip())
+                        leechers = int(res.find('td', {'class': 'table_leechers'}).find('a').text.strip())
+                        entry['torrent_seeds'] = seeds
+                        entry['torrent_leeches'] = leechers
+
+                        size = res.find('td', attrs={'class': re.compile('table_size')}).text
+                        size = re.search('(\d+(?:[.,]\d+)*)\s?([KMG]B)', size)
+
+                        if size:
+                            if size.group(2) == 'GB':
```
Comment (cvium) [inline flexget/plugins/sites/freshon.py]: https://github.com/Flexget/Flexget/blob/develop/flexget/utils/tools.py#L437

File: flexget/plugins/sites/freshon.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+                                entry['content_size'] = int(float(size.group(1)) * 1000 ** 2 / 1024 ** 2)
+                            elif size.group(2) == 'KB':
+                                entry['content_size'] = int(float(size.group(1)) * 1000 / 1024 ** 2)
+                            else:
+                                entry['content_size'] = int(float(size.group(1)) / 1024 ** 2)
+
+                        if (int(entry['torrent_seeds']) > 0):
+                            entries.add(entry)
+                        else:
+                            log.debug('0 SEEDS, not adding entry')
```
Comment (cvium) [inline flexget/plugins/sites/freshon.py]: I'm a little puzzled by this and I assume it was always there. That should not be up to the search plugin decide. We already have https://flexget.com/Plugins/torrent_alive if people don't want dead torrents.

File: flexget/plugins/sites/freshon.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+            'passkey': {'type': 'string'},
+            'category': {
+                'type': 'string',
+                'default': 'all',
+                'enum': ['hd', 'webdl', 'all']
+            },
+            'freeleech': {
+                'type': 'string',
+                'default': 'all',
+                'enum': ['free', 'half', 'all']
```
Comment (liiight) [inline flexget/plugins/sites/freshon.py]: Best to use this as a dict for reusability:
```python
LEECHSTATUS = {
    'all': 0,
    'free': 3,
    'half': 4
}
```
then in the schema use:
```python
'enum': list(LEECHSTATUS.keys())
```
and later you can access this dict too

File: flexget/plugins/sites/freshon.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        login_fragment = 'login.php?action=makelogin'
+
+        if not task.requests.cookies:
+            log.debug('Logging in to Freshon.tv...')
+            params = {'username': config['username'],
+                      'password': config['password'],
+                      'login': 'Do it!'}
+            urlstr = (base_url, login_fragment)
+            lsrc = task.requests.post('/'.join(urlstr), data=params)
+            if str(config['username']) in lsrc.text:
```
Comment (liiight) [inline flexget/plugins/sites/freshon.py]: no need for `str`

File: flexget/plugins/sites/freshon.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+                log.debug('Login to FreshonTV was successful')
+            elif str('Username does not exist in the userbase') in lsrc.text:
+                log.error('Invalid username or password for FreshonTV; Check your settings')
+                raise plugin.PluginError("Invalid username or Password for FreshonTV.")
+            else:
+                log.error('Login to FreshonTV was NOT successful')
+                raise plugin.PluginError("Login to FreshonTV was NOT successful")
+
+        if config['freeleech'] == 'all':
```
Comment (liiight) [inline flexget/plugins/sites/freshon.py]: See above about reusing dicts

File: flexget/plugins/sites/freshon.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        elif config['freeleech'] == 'free':
+            freeleech = 3
+        elif config['freeleech'] == 'half':
+            freeleech = 4
+
+        entries = set()
+        for search_string in entry.get('search_strings', [entry['title']]):
+            query = normalize_unicode(search_string.replace(' ', '+'))
+            query_url_fragment = urllib.quote_plus(query.encode('utf-8'))
+            str_url = (base_url, search_fragment)
```
Comment (liiight) [inline flexget/plugins/sites/freshon.py]: no need for `str`, separate urls and params

File: flexget/plugins/sites/freshon.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+            url = '/'.join(str_url) % (query_url_fragment, config['category'], str(freeleech))
+            log.debug('Search url: %s' % url)
+            page = task.requests.get(url).content
```
Comment (liiight) [inline flexget/plugins/sites/freshon.py]: best to wrap this in a try, except with `RequestsException`

File: flexget/plugins/sites/freshon.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+                # limit to 10 pages
+                if max_page_number > 10:
+                    max_page_number = 10
+
+                # pages are 0 based
+                nextpage = 0
+                while (nextpage < max_page_number):
+                    if (nextpage > 0):
+                        newurl = url + '&page=' + str(nextpage)
+                        log.debug('-----> NEXT PAGE : %s' % newurl)
```
Comment (liiight) [inline flexget/plugins/sites/freshon.py]: pass string interpolation to logger, replace `%` with `,`:
` log.debug('-----> NEXT PAGE : %s', newurl)`

File: flexget/plugins/sites/freshon.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+                        f1 = task.requests.get(newurl).content
+                        soup = get_soup(f1)
+                    for res in soup.findAll('tr', {'class': re.compile('torrent_[0-9]*')}):
+                        entry = Entry()
+                        # skip if nuked
+                        if res.find('img', alt='Nuked') != None:
```
Comment (liiight) [inline flexget/plugins/sites/freshon.py]: maybe add log for this scenario?

File: flexget/plugins/sites/freshon.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+                        else:
+                            log.debug('0 SEEDS, not adding entry')
+                    nextpage += 1
+
+        return entries
+
+
+@event('plugin.register')
+def register_plugin():
+    plugin.register(SearchFreshon, 'freshon', groups=['search'], api_ver=2)
```
Comment (liiight) [inline flexget/plugins/sites/freshon.py]: change `groups` to `interfaces`

**Comment by Jesus-21** [comment]
It needed a slight review indeed, I cleaned it and made sure it was PEP8 compliant in the process.
I also added `page_limit` to the config, hardcoding a limit to 10 pages seemed a bad idea to me.

File: flexget/plugins/sites/freshon.py [hunk 11]
```diff
@@ -9,10 +9,21 @@
...
 from flexget.utils.soup import get_soup
+from flexget.utils.tools import parse_filesize
 from flexget.utils.search import normalize_unicode
 
 log = logging.getLogger('search_freshon')
 
+BASE_URL = 'https://freshon.tv'
+SEARCH_FRAGMENT = 'browse.php?search=%s&tab=%s&incldead=%s'
+DL_FRAGMENT = 'download.php?type=rss&id=%s&passkey=%s'
+LOGIN_FRAGMENT = 'login.php?action=makelogin'
```
Comment (cvium) [inline flexget/plugins/sites/freshon.py]: Can't all these "fragments" be replaced with dicts instead? Like you're doing in `login`.

File: flexget/plugins/sites/freshon.py [hunk 12]
```diff
@@ -176,8 +181,13 @@ def parse_entry(self, res):
 
         details_url = res.find('a', {'class': 'torrent_name_link'})['href']
         id = int((re.match('.*?([0-9]+)$', details_url).group(1)).strip())
-        url = (BASE_URL, DL_FRAGMENT)
-        entry['url'] = '/'.join(url) % (id, self.config['passkey'])
+        params = {
+            'type': 'rss',
+            'id': id,
```
Comment (cvium) [inline flexget/plugins/sites/freshon.py]: `id` is a python keyword. Please call it something else.

**Comment by liiight** [comment]
Looks good to me

**Pull request merged**